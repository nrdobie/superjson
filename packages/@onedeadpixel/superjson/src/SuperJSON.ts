/**
 * @module @onedeadpixel/superjson
 */
import isObject from 'isobject'
import { compose } from './compose'

type KlassType<Klass = any> = new (...args: any[]) => Klass

/**
 * Options for SuperJSON
 */
interface ISuperJSONOptions {
  /**
   * Special key to signify a SuperJSON serialized type in stringified code.
   */
  typeKey: string

  /**
   * Special key to signify a SuperJSON serialized value in stringified code.
   */
  valueKey: string
}

/**
 * SuperJSON transformer
 */
interface ISuperJSONTransform<Klass = any> {
  /**
   * Class for transformer
   */
  klass: KlassType<Klass>

  /**
   * Name of transformer type
   */
  name: string

  /**
   * Function to turn instance into JSON value
   */
  toJSONValue: (instance: Klass) => any

  /**
   * Function to turn JSON value into instance
   */
  fromJSONValue: (value: any) => Klass
}

interface ISuperJSONRegisterOptions<Klass> {
  /**
   * Name of transformer type
   */
  name?: string

  /**
   * Function to turn instance into JSON value
   */
  toJSONValue?: (instance: Klass) => any

  /**
   * Function to turn JSON value into instance
   */
  fromJSONValue?: (value: any) => Klass
}

/**
 * @hidden
 */
const originalJSON = global.JSON

export class SuperJSON {
  /**
   * Standalone reviver function that can be passed to `JSON.parse`
   */
  public get reviver() {
    const self = this
    return function sjsonReviver(key: any, value: any) {
      return self.jsonReviver(key, value)
    }
  }

  /**
   * Standalone reviver function that can be passed to `JSON.parse`
   */
  public get replacer() {
    const self = this
    return function sjsonReplacer(key: any, value: any) {
      return self.jsonReplacer(key, value, this[key])
    }
  }

  /**
   * Default options for SuperJSON
   */
  public static DEFAULT_OPTIONS: ISuperJSONOptions = {
    typeKey: '__sj_type',
    valueKey: '__sj_value'
  }

  /**
   * Settings for SuperJSON instance
   */
  public readonly options: ISuperJSONOptions

  /**
   * Active transforms
   */
  private registeredTransforms: Map<string, ISuperJSONTransform> = new Map()

  /**
   * Lookup table to convert Class into name
   */
  private klassLookup: Map<KlassType, string> = new Map()

  constructor(options: Partial<ISuperJSONOptions> = {}) {
    this.options = {
      ...options,
      ...SuperJSON.DEFAULT_OPTIONS
    }
  }

  /**
   * `JSON.parse` compatible function to convert JSON string into JavaScript objects
   *
   * @param text JSON string to convert to JavaScript objects
   * @param reviver If a function, this prescribes how the value originally produced by parsing is transformed, before being returned.
   */
  public parse(text: string, reviver?: (key: any, value: any) => any): any {
    return originalJSON.parse(
      text,
      compose(
        this.reviver,
        reviver
      )
    )
  }

  /**
   * `JSON.stringify` compatible function to convert JavaScript objects into JSON string
   *
   * @param value The value to convert to a JSON string.
   * @param replacer A function that alters the behavior of the stringification process, or an array of String and Number objects that serve as a whitelist for selecting/filtering the properties of the value object to be included in the JSON string. If this value is null or not provided, all properties of the object are included in the resulting JSON string.
   * @param space A String or Number object that's used to insert white space into the output JSON string for readability purposes.
   */
  public stringify(
    value: any,
    replacer?: (key: any, value: any) => any,
    space?: string | number
  ) {
    return originalJSON.stringify(
      value,
      compose(
        this.replacer,
        replacer
      ),
      space
    )
  }

  /**
   * Adds a Class to be serialized by SuperJSON
   *
   * @param klass Class to add to SuperJSON for serialization
   * @param options Additional settings for controlling serialization
   */
  public register<Klass>(
    klass: KlassType<Klass>,
    options: ISuperJSONRegisterOptions<Klass> = {}
  ) {
    let name: string
    if (options.name) {
      name = options.name
    } else if ('name' in klass) {
      name = klass.name
    } else {
      throw new TypeError(`name must be provided by options or class`)
    }

    let toJSONValue: ISuperJSONTransform<Klass>['toJSONValue']
    if (options.toJSONValue) {
      toJSONValue = options.toJSONValue
    } else if ('toJSONValue' in klass.prototype) {
      toJSONValue = (instance: any) => instance.toJSONValue()
    } else {
      throw new TypeError(`toJSONValue must be provided by options or class`)
    }

    let fromJSONValue: ISuperJSONTransform<Klass>['fromJSONValue']
    if (options.fromJSONValue) {
      fromJSONValue = options.fromJSONValue
    } else if ('fromJSONValue' in klass) {
      fromJSONValue = value => (klass as any).fromJSONValue(value)
    } else {
      throw new TypeError(`fromJSONValue must be provided by options or class`)
    }

    const transform: ISuperJSONTransform<Klass> = {
      klass,
      name,
      toJSONValue,
      fromJSONValue
    }

    this.registeredTransforms.set(name, transform)
    this.klassLookup.set(klass, name)

    return this
  }

  /**
   * Removes a Class from being serialized by SuperJSON
   *
   * @param klass Class to remove from SuperJSON
   */
  public unregister(klass: new (...args: any[]) => any): this

  /**
   * Removes a Class from being serialized by SuperJSON
   *
   * @param name Name of Class to remove from SuperJSON
   */
  public unregister(name: string): this
  public unregister(klassOrName: (new (...args: any[]) => any) | string) {
    let name: string
    if (typeof klassOrName !== 'string') {
      name = this.klassLookup.get(klassOrName)
    } else {
      name = klassOrName
    }

    if (!this.registeredTransforms.has(name)) {
      throw new Error(`${name || (klassOrName as any).name} is not registered`)
    }

    const transform = this.registeredTransforms.get(name)

    this.registeredTransforms.delete(name)
    this.klassLookup.delete(transform.klass)

    return this
  }

  /**
   * Gets the transformer for a Class
   *
   * @param instance instance to get transformer for
   */
  public getTransformer(instance: {
    constructor: KlassType
  }): ISuperJSONTransform

  /**
   * Gets the transformer for a Class
   *
   * @param klass Class to get transformer for
   */
  public getTransformer(klass: new (...args: any[]) => any): ISuperJSONTransform

  /**
   * Gets the transformer for a Class
   *
   * @param name name to get transformer for
   */
  public getTransformer(name: string): ISuperJSONTransform
  public getTransformer(
    instanceKlassOrName: { constructor: KlassType } | KlassType | string
  ) {
    let name: string
    if (typeof instanceKlassOrName !== 'string') {
      let klass: any
      if (instanceKlassOrName.constructor === Function) {
        klass = instanceKlassOrName
      } else {
        klass = instanceKlassOrName.constructor
      }

      name = this.klassLookup.get(klass)

      if (!name) {
        throw Error(`${klass.name} is not registered`)
      }
    } else {
      name = instanceKlassOrName
    }

    return this.registeredTransforms.get(name)
  }

  /**
   * Used as a `JSON.parse` reviver to convert JSON values into objects.
   *
   * @param key Object key
   * @param value Object value
   */
  private jsonReviver(key: any, value: any) {
    if (this.isRevivable(value)) {
      const type = value[this.options.typeKey]
      const jsonValue = value[this.options.valueKey]
      const transform = this.getTransformer(type)

      return transform.fromJSONValue(jsonValue)
    }

    return value
  }

  /**
   * Checks if the value has a transformer
   *
   * @param value value to test
   */
  private isRevivable(value: any) {
    return (
      isObject(value) &&
      this.options.typeKey in value &&
      this.options.valueKey in value &&
      this.registeredTransforms.has(value[this.options.typeKey])
    )
  }

  /**
   * Used as a `JSON.stringify` replacer to convert objects into JSON values.
   *
   * @param key Object key
   * @param value Object value
   * @param originalValue Object value before `.toJSON` was called.
   */
  private jsonReplacer(key: any, value: any, originalValue: any) {
    const transformValue =
      value !== originalValue && value !== undefined ? originalValue : value

    if (this.isReplaceable(transformValue)) {
      const transformer = this.getTransformer(transformValue)

      const jsonValue = transformer.toJSONValue(transformValue)

      return {
        [this.options.typeKey]: transformer.name,
        [this.options.valueKey]: jsonValue
      }
    }

    return value
  }

  /**
   * Checks if the value has a transformer
   *
   * @param value value to test
   */
  private isReplaceable(value: any) {
    return (
      isObject(value) &&
      'constructor' in value &&
      this.klassLookup.has(value.constructor)
    )
  }
}
