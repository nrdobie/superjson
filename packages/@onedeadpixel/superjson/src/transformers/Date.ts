/**
 * @module @onedeadpixel/superjson
 */
import { SuperJSON } from '../SuperJSON'

/**
 * Registers the core JS Date class
 *
 * @param sjson SuperJSON instance to register Date to
 */
export function registerDate(sjson: SuperJSON) {
  /* istanbul ignore else */
  if (Date) {
    sjson.register(Date, {
      name: 'Date',
      toJSONValue: date => date.toISOString(),
      fromJSONValue: isoString => new Date(isoString)
    })
  }
}
