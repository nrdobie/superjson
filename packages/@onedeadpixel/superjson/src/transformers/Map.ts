/**
 * @module @onedeadpixel/superjson
 */
import { SuperJSON } from '../SuperJSON'

/**
 * Registers the core JS Map class
 *
 * @param sjson SuperJSON instance to register Map to
 */
export function registerMap(sjson: SuperJSON) {
  /* istanbul ignore else */
  if (Map) {
    sjson.register(Map, {
      name: 'Map',
      toJSONValue: map => Array.from(map.entries()),
      fromJSONValue: values => new Map(values)
    })
  }
}
