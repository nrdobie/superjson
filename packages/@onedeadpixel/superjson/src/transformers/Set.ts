/**
 * @module @onedeadpixel/superjson
 */
import { SuperJSON } from '../SuperJSON'

/**
 * Registers the core JS Set class
 *
 * @param sjson SuperJSON instance to register Set to
 */
export function registerSet(sjson: SuperJSON) {
  /* istanbul ignore else */
  if (Set) {
    sjson.register(Set, {
      name: 'Set',
      toJSONValue: set => Array.from(set.values()),
      fromJSONValue: values => new Set(values)
    })
  }
}
