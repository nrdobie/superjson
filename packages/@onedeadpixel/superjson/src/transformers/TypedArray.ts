/**
 * @module @onedeadpixel/superjson
 */
import { SuperJSON } from '../SuperJSON'

/**
 * Map of TypedArray constructors
 */
const typedArrayConstructors = new Map<string, any>([
  ['Int8Array', Int8Array],
  ['Uint8Array', Uint8Array],
  ['Uint8ClampedArray', Uint8ClampedArray],
  ['Int16Array', Int16Array],
  ['Uint16Array', Uint16Array],
  ['Int32Array', Int32Array],
  ['Uint32Array', Uint32Array],
  ['Float32Array', Float32Array],
  ['Float64Array', Float64Array]
])

/**
 * Registers the core JS TypedArrays
 *
 * Supports the following TypedArrays:
 *
 *  - Int8Array
 *  - Uint8Array
 *  - Uint8ClampedArray
 *  - Int16Array
 *  - Uint16Array
 *  - Int32Array
 *  - Uint32Array
 *  - Float32Array
 *  - Float64Array
 *
 * @param sjson SuperJSON instance to register TypedArrays to
 */
export function registerTypedArrays(sjson: SuperJSON) {
  for (const [name, TypedArray] of typedArrayConstructors.entries()) {
    /* istanbul ignore else */
    if (TypedArray) {
      sjson.register(TypedArray, {
        name,
        toJSONValue: (typedArray: any) => Array.from(typedArray.values()),
        fromJSONValue: arrayOfValues => TypedArray.from(arrayOfValues)
      })
    }
  }
}
