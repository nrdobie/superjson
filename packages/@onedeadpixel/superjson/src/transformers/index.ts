/**
 * @module @onedeadpixel/superjson
 */
import { SuperJSON } from '../SuperJSON'
import { registerDate } from './Date'
import { registerMap } from './Map'
import { registerSet } from './Set'
import { registerTypedArrays } from './TypedArray'

export * from './Date'
export * from './Map'
export * from './Set'
export * from './TypedArray'

/**
 * Registers all the core JS classes
 *
 * @param sjson SuperJSON instance to register classes to
 */
export function registerAll(sjson: SuperJSON) {
  registerDate(sjson)
  registerMap(sjson)
  registerSet(sjson)
  registerTypedArrays(sjson)
}
