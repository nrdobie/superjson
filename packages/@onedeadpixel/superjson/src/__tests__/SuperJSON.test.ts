import { SuperJSON } from '../SuperJSON'
import { registerDate, registerMap } from '../transformers'

class TestAnimal {
  public static fromJSONValue(value: any) {
    return new TestAnimal(value.name)
  }

  constructor(public name: string) {}

  public toJSONValue() {
    return { name: this.name }
  }
}

const TestFail = { prototype: {} }

describe('SuperJSON', () => {
  let sjson: SuperJSON
  beforeEach(() => {
    sjson = new SuperJSON()

    registerDate(sjson)
  })

  describe('.reviver', () => {
    it('should be usable standalone', () => {
      const testString = `{"test": { "__sj_type": "Date", "__sj_value": "1995-12-04T00:00:00Z" } }`

      expect(JSON.parse(testString, sjson.reviver)).toEqual({
        test: new Date('1995-12-04T00:00:00Z')
      })
    })
  })

  describe('.replacer', () => {
    it('should be usable standalone', () => {
      expect(
        JSON.stringify(
          { test: new Date('1995-12-04T00:00:00Z') },
          sjson.replacer
        )
      ).toMatchSnapshot()
    })
  })

  describe('#constructor', () => {
    it('should allow for custom keys', () => {
      sjson = new SuperJSON({ typeKey: '__diff_key', valueKey: '__diff_value' })
      registerDate(sjson)

      const str = sjson.stringify(new Date('1995-12-04T00:00:00Z'))
      expect(str).toMatchSnapshot()
      const parsed = sjson.parse(str)
      expect(parsed).toEqual(new Date('1995-12-04T00:00:00Z'))
    })
  })

  describe('#parse', () => {
    it('should parse a regular JSON string', () => {
      const testString = `{"test": true}`

      expect(sjson.parse(testString)).toEqual({ test: true })
    })

    it('should parse a SuperJSON string', () => {
      const testString = `{"test": { "__sj_type": "Date", "__sj_value": "1995-12-04T00:00:00Z" } }`

      expect(sjson.parse(testString)).toEqual({
        test: new Date('1995-12-04T00:00:00Z')
      })
    })

    it('should support custom reviver', () => {
      const testString = `{"testOne": { "__sj_type": "Date", "__sj_value": "1995-12-04T00:00:00Z" }, "testTwo": true }`

      const reviver = (key: any, value: any) =>
        key === 'testTwo' ? undefined : value

      expect(sjson.parse(testString, reviver)).toEqual({
        testOne: new Date('1995-12-04T00:00:00Z')
      })
    })
  })

  describe('#stringify', () => {
    it('should stringify regular object', () => {
      const testObject = { test: true }

      expect(sjson.stringify(testObject)).toMatchSnapshot()
    })

    it('should stringify a SuperJSON object', () => {
      const testObject = { test: new Date('1995-12-04T00:00:00Z') }

      expect(sjson.stringify(testObject)).toMatchSnapshot()
    })

    it('should support custom replacer', () => {
      const testString = {
        testOne: new Date('1995-12-04T00:00:00Z'),
        testTwo: true
      }

      const replacer = (key: any, value: any) =>
        key === 'testTwo' ? undefined : value

      expect(sjson.stringify(testString, replacer)).toMatchSnapshot()
    })
  })

  describe('#register', () => {
    let testAnimal: TestAnimal
    beforeEach(() => {
      testAnimal = new TestAnimal('Odin')
    })

    it('should be able to use an instance', () => {
      sjson.register(TestAnimal)

      const str = sjson.stringify(testAnimal)
      expect(str).toMatchSnapshot()
      const parsed = sjson.parse(str)
      expect(parsed.name).toBe('Odin')
    })

    it('should be able to override name', () => {
      sjson.register(TestAnimal, { name: 'TesterAnimal' })

      const str = sjson.stringify(testAnimal)
      expect(str).toMatchSnapshot()
      const parsed = sjson.parse(str)
      expect(parsed.name).toBe('Odin')
    })

    it('should be able to override toJSONValue and fromJSONValue', () => {
      sjson.register(TestAnimal, {
        toJSONValue: animal => ({ betterName: animal.name }),
        fromJSONValue: value => new TestAnimal(value.betterName)
      })

      const str = sjson.stringify(testAnimal)
      expect(str).toMatchSnapshot()
      const parsed = sjson.parse(str)
      expect(parsed.name).toBe('Odin')
    })

    it('should require a name', () => {
      expect(() =>
        sjson.register(TestFail as any, {
          toJSONValue: () => ({}),
          fromJSONValue: () => ({})
        })
      ).toThrowErrorMatchingSnapshot()
    })

    it('should require a toJSONValue', () => {
      expect(() =>
        sjson.register(TestFail as any, {
          name: 'TestFail',
          fromJSONValue: () => ({})
        })
      ).toThrowErrorMatchingSnapshot()
    })

    it('should require a fromJSONValue', () => {
      expect(() =>
        sjson.register(TestFail as any, {
          name: 'TestFail',
          toJSONValue: () => ({})
        })
      ).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#unregister', () => {
    it('should unregister by class', () => {
      registerMap(sjson)

      sjson.unregister(Map)

      expect(sjson.stringify(new Map([[1, 'target']]))).toMatchSnapshot()
    })

    it('should unregister by name', () => {
      registerMap(sjson)

      sjson.unregister('Map')

      expect(sjson.stringify(new Map([[1, 'target']]))).toMatchSnapshot()
    })

    it('should throw error for unregistered type', () => {
      expect(() => sjson.unregister(Map)).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#getTransformer', () => {
    it('should reverse a string', () => {
      const transformer = sjson.getTransformer('Date')

      expect(transformer.klass).toBe(Date)
    })

    it('should reverse a class', () => {
      const transformer = sjson.getTransformer(Date)

      expect(transformer.klass).toBe(Date)
    })

    it('should reverse an instance', () => {
      const date = new Date()
      const transformer = sjson.getTransformer(date as any)

      expect(transformer.klass).toBe(Date)
    })

    it('should throw error on unregistered type', () => {
      expect(() => sjson.getTransformer(Map)).toThrowErrorMatchingSnapshot()
    })
  })
})
