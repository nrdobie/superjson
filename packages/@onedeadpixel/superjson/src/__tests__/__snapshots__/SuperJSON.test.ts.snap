// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuperJSON #constructor should allow for custom keys 1`] = `"{\\"__sj_type\\":\\"Date\\",\\"__sj_value\\":\\"1995-12-04T00:00:00.000Z\\"}"`;

exports[`SuperJSON #getTransformer should throw error on unregistered type 1`] = `"Map is not registered"`;

exports[`SuperJSON #register should be able to override name 1`] = `"{\\"__sj_type\\":\\"TesterAnimal\\",\\"__sj_value\\":{\\"name\\":\\"Odin\\"}}"`;

exports[`SuperJSON #register should be able to override toJSONValue and fromJSONValue 1`] = `"{\\"__sj_type\\":\\"TestAnimal\\",\\"__sj_value\\":{\\"betterName\\":\\"Odin\\"}}"`;

exports[`SuperJSON #register should be able to use an instance 1`] = `"{\\"__sj_type\\":\\"TestAnimal\\",\\"__sj_value\\":{\\"name\\":\\"Odin\\"}}"`;

exports[`SuperJSON #register should require a fromJSONValue 1`] = `"fromJSONValue must be provided by options or class"`;

exports[`SuperJSON #register should require a name 1`] = `"name must be provided by options or class"`;

exports[`SuperJSON #register should require a toJSONValue 1`] = `"toJSONValue must be provided by options or class"`;

exports[`SuperJSON #stringify should stringify a SuperJSON object 1`] = `"{\\"test\\":{\\"__sj_type\\":\\"Date\\",\\"__sj_value\\":\\"1995-12-04T00:00:00.000Z\\"}}"`;

exports[`SuperJSON #stringify should stringify regular object 1`] = `"{\\"test\\":true}"`;

exports[`SuperJSON #stringify should support custom replacer 1`] = `"{\\"testOne\\":{\\"__sj_type\\":\\"Date\\",\\"__sj_value\\":\\"1995-12-04T00:00:00.000Z\\"}}"`;

exports[`SuperJSON #unregister should throw error for unregistered type 1`] = `"Map is not registered"`;

exports[`SuperJSON #unregister should unregister by class 1`] = `"{}"`;

exports[`SuperJSON #unregister should unregister by name 1`] = `"{}"`;

exports[`SuperJSON .replacer should be usable standalone 1`] = `"{\\"test\\":{\\"__sj_type\\":\\"Date\\",\\"__sj_value\\":\\"1995-12-04T00:00:00.000Z\\"}}"`;
